storage-proofs-porep/src/stacked/vanilla/cores.rs

```diff
@@ -176,18 +177,22 @@ fn create_core_units(
     let core_units = (0..unit_count)
         .map(|i| {
             (0..cores_per_unit)
-                .map(|j| {
+                .filter_map(|j| {
                     // Every group gets a single unit assigned first. Only if all groups have
                     // already one unit, a second one will be assigned if possible. This would then
                     // be the second "round" of assignments.
                     let round = i / group_count;
                     // The index of the core that is bound to a unit.
                     let core_index = (j + i * group_size) % core_count + (round * cores_per_unit);
                     assert!(core_index < core_count);
-                    core_index
+
+                    allowed_cores
+                        .is_set(core_index.try_into().ok()?)
+                        .then_some(core_index)
                 })
                 .collect::<Vec<_>>()
         })
+        .filter(|x| !x.is_empty())
```

Comment:
This only filters out if a whole "group" is empty, right? Could you please add  a test case (or modify an existing one), when the allowed cores reduce the number of members of a group, but don't leave it completely empty? Or is this this something that never happens in reality?

Reply:
Correct, this removes empty vecs from the top level which may exist after the filtering is done.

Test added: https://github.com/filecoin-project/rust-fil-proofs/blob/ae4019f2405a4acb21964e5027a6406c070b42eb/storage-proofs-porep/src/stacked/vanilla/cores.rs#L422-L433